yargs.js

```diff
@@ -669,6 +670,42 @@ function Yargs (processArgs, cwd, parentRequire) {
     return detectLocale
   }
 
+  var exitFn = process.exit
+  var exiting = false
+  self.exit = function (fn) {
+    // if a custom exit handler is provided we emit
+    // the event on the next tick of the event loop, giving
+    // the handler access to the parsed argv.
+    exitFn = function (code) {
+      if (exiting) return
+      exiting = true
+      process.nextTick(function () {
```

Comment:
If validation fails, the following will execute (when they wouldn't before) because this is asynchronous:

1. a command handler
2. any top-level logic that runs after parsing